### PR TITLE
feat(processor): enhance reports to hold transformation and tracking plan metrics

### PR DESCRIPTION
**Pull request by Jayachand**
# Description

> Enhancing reports table to hold transformation and tracking plan information
> New connectionDetails columns
- TransformationId
- TrannsformationVersionId
- TrackingPlanId
- TrackingPlanVersion
> New statusDetails columns
- ViolationCount
- ErrorType

## Notion Ticket

https://www.notion.so/rudderstacks/Send-violation-information-tp-id-version-violation-message-payload-c213bd096b1744189ce570d33026cdd2

https://www.notion.so/rudderstacks/Send-Transformation-Id-information-in-report-metrics-912964eb39144675b8905639e471c8a8?pvs=4

## Security

- [ ] The code changed/added as part of this pull request won't create any security issues with how the software is being used.

### CODE REVIEW

File: sql/migrations/reports/000004_alter_reports_add_tr_tp_columns.up.sql [hunk 1]
```diff
@@ -0,0 +1,7 @@
+ALTER TABLE reports
+ADD COLUMN IF NOT EXISTS transformation_id TEXT DEFAULT '',
+ADD COLUMN IF NOT EXISTS transformation_version_id TEXT DEFAULT '',
```
Comment (satishrudderstack) [inline sql/migrations/reports/000004_alter_reports_add_tr_tp_columns.up.sql:3]: @Jayachand Shouldn't this be `INT` like for tracking-plan ?

File: utils/types/reporting_types.go [hunk 2]
```diff
@@ -35,6 +35,14 @@ var (
 	WAREHOUSE              = "warehouse"
 )
 
+var (
+	// Tracking plan validation states
+	SUCCEEDED                   = "succeeded"
+	SUCCEEDED_WITHOUT_VIOLATION = "succeeded_without_violation"
+	SUCCEEDED_WITH_VIOLATION    = "succeeded_with_violation"
```
[the remark this answers is not in the document]
Comment (Jayachand) [reply utils/types/reporting_types.go]: done

File: sql/migrations/reports/000004_alter_reports_add_tr_tp_columns.up.sql (hunk 1 above)
Comment (Jayachand) [reply sql/migrations/reports/000004_alter_reports_add_tr_tp_columns.up.sql:3]: transformation_version_id is a 32 character string whereas trackingplan version is a number

File: enterprise/reporting/reporting.go [hunk 3]
```diff
@@ -244,7 +267,22 @@ func (*HandleT) getAggregatedReports(reports []*types.ReportByStatus) []*types.M
...
+			report.ConnectionDetails.SourceTaskRunID,
+			report.ConnectionDetails.SourceJobID,
+			report.ConnectionDetails.SourceJobRunID,
+			report.ConnectionDetails.TransformationId,
+			report.ConnectionDetails.TransformationVersionId,
+			report.ConnectionDetails.TrackingPlanId,
+			fmt.Sprint(report.ConnectionDetails.TrackingPlanVersion),
+			report.PUDetails.InPU, report.PUDetails.PU,
+			report.StatusDetail.Status,
+			fmt.Sprint(report.StatusDetail.StatusCode),
```
Comment (Sidddddarth) [inline enterprise/reporting/reporting.go:283]: same as above

File: enterprise/reporting/reporting.go [hunk 4]
```diff
@@ -409,7 +454,7 @@ func (r *HandleT) sendMetric(ctx context.Context, netClient *http.Client, client
 		httpRequestStart := time.Now()
 		resp, err := netClient.Do(req)
 		if err != nil {
-			r.log.Error(err.Error())
+			// r.log.Error(err.Error())
```
Comment (Sidddddarth) [inline enterprise/reporting/reporting.go:457]: Any reason why this is commented

File: enterprise/reporting/reporting.go [hunk 5]
```diff
@@ -433,7 +478,7 @@ func (r *HandleT) sendMetric(ctx context.Context, netClient *http.Client, client
 
 	b := backoff.WithContext(backoff.NewExponentialBackOff(), ctx)
 	err = backoff.RetryNotify(operation, b, func(err error, t time.Duration) {
-		r.log.Errorf(`[ Reporting ]: Error reporting to service: %v`, err)
+		// r.log.Errorf(`[ Reporting ]: Error reporting to service: %v`, err)
```
Comment (Sidddddarth) [inline enterprise/reporting/reporting.go:481]: Any reason why this is commented

File: router/router.go [hunk 6]
```diff
@@ -1323,7 +1323,7 @@ func (rt *HandleT) commitStatusList(responseList *[]jobResponseT) {
 			if rt.transientSources.Apply(parameters.SourceID) {
 				sampleEvent = routerutils.EmptyPayload
 			}
-			sd = utilTypes.CreateStatusDetail(resp.status.JobState, 0, errorCode, string(resp.status.ErrorResponse), sampleEvent, eventName, eventType)
+			sd = utilTypes.CreateStatusDetail(resp.status.JobState, 0, 0, errorCode, string(resp.status.ErrorResponse), sampleEvent, eventName, eventType, "")
```
Comment (Sidddddarth) [inline router/router.go:1326]: we might as well use the `errorType` for errors at router.

**Review by lvrach — APPROVED** [review]
Approved with minor comments

File: router/batchrouter/batchrouter.go [hunk 7]
```diff
@@ -1324,7 +1324,6 @@ func (brt *HandleT) setJobStatus(batchJobs *BatchJobsT, isWarehouse bool, errOcc
 
 	// REPORTING - START
 	if brt.reporting != nil && brt.reportingEnabled {
-		types.AssertSameKeys(connectionDetailsMap, statusDetailsMap)
```
Comment (Sidddddarth) [inline router/batchrouter/batchrouter.go:1327]: keeping this would've helped validate these changes.

File: processor/processor.go [hunk 8]
```diff
@@ -910,17 +914,41 @@ func (proc *Handle) recordEventDeliveryStatus(jobsByDestID map[string][]*jobsdb.
...
+		for _, message := range messages {
+			proc.updateMetricMaps(successCountMetadataMap, successCountMap, connectionDetailsMap, statusDetailsMap, userTransformedEvent, jobsdb.Succeeded.State, stage, func() json.RawMessage {
+				if stage != transformer.TrackingPlanValidationStage {
+					return []byte(`{}`)
+				}
+				sampleEvent, err := jsonfast.Marshal(message)
+				if err != nil {
+					proc.logger.Errorf(`[Processor: getDestTransformerEvents] Failed to unmarshal first element in transformed events: %v`, err)
+				}
+				if err != nil || proc.transientSources.Apply(commonMetaData.SourceID) {
```
Comment (Sidddddarth) [inline processor/processor.go:946]: Transient sources check should be held at the top priority and the sample event should be empty in case its a transient source

File: processor/trackingplan.go [hunk 9]
```diff
@@ -102,7 +102,7 @@ func (proc *Handle) validateEvents(groupedEventsByWriteKey map[WriteKeyT][]trans
 		trackingPlanEnabledMap[SourceIDT(sourceID)] = true
 
 		var successMetrics []*types.PUReportedMetric
-		eventsToTransform, successMetrics, _, _ := proc.getDestTransformerEvents(response, commonMetaData, destination, transformer.TrackingPlanValidationStage, true, false) // Note: Sending false for usertransformation enabled is safe because this stage is before user transformation.
+		eventsToTransform, successMetrics, _, _ := proc.getDestTransformerEvents(response, commonMetaData, eventsByMessageID, destination, transformer.TrackingPlanValidationStage, true, false) // Note: Sending false for usertransformation enabled is safe because this stage is before user transformation.
```
Comment (Sidddddarth) [inline processor/trackingplan.go:105]: Consider using another function for tracking plan without disturbing the flow for userTransform and destTransformation

File: enterprise/reporting/reporting.go (hunk 4 above)
Comment (Jayachand) [reply enterprise/reporting/reporting.go:457]: omitted

File: enterprise/reporting/reporting.go (hunk 5 above)
Comment (Jayachand) [reply enterprise/reporting/reporting.go:481]: deleted

File: enterprise/reporting/reporting.go [hunk 10]
```diff
@@ -244,7 +267,22 @@ func (*HandleT) getAggregatedReports(reports []*types.ReportByStatus) []*types.M
...
+			report.InstanceDetails.WorkspaceID, report.InstanceDetails.Namespace, report.InstanceDetails.InstanceID,
+			report.ConnectionDetails.SourceID,
+			report.ConnectionDetails.DestinationID,
+			report.ConnectionDetails.SourceTaskRunID,
+			report.ConnectionDetails.SourceJobID,
+			report.ConnectionDetails.SourceJobRunID,
+			report.ConnectionDetails.TransformationId,
+			report.ConnectionDetails.TransformationVersionId,
+			report.ConnectionDetails.TrackingPlanId,
+			fmt.Sprint(report.ConnectionDetails.TrackingPlanVersion),
```
[the remark this answers is not in the document]
Comment (Jayachand) [reply enterprise/reporting/reporting.go]: done

File: processor/trackingplan.go (hunk 9 above)
Comment (Jayachand) [reply processor/trackingplan.go:105]: I think we should start showing sample event for success state in userTransform and destTransformation too. Is there any reason behind not showing it. 
From reporting side, this might increase computation while generating hourly view data. they have asked to hold sending this data until timescale upgrade task is done. For this reason I have used the same existing function.

File: utils/types/reporting_types.go [hunk 11]
```diff
@@ -85,6 +95,10 @@ type ConnectionDetails struct {
 	SourceDefinitionId      string `json:"sourceDefinitionId"`
 	DestinationDefinitionId string `string:"destinationDefinitionId"`
 	SourceCategory          string `json:"sourceCategory"`
+	TransformationId        string `json:"transformationId"`
+	TransformationVersionId string `json:"transformationVersionId"`
+	TrackingPlanId          string `json:"trackingPlanId"`
+	TrackingPlanVersion     int    `json:"trackingPlanVersion"`
```
[the remark this answers is not in the document]
Comment (Jayachand) [reply utils/types/reporting_types.go]: converted Id to ID format for newly added properties.

File: processor/processor.go (hunk 8 above)
Comment (Jayachand) [reply processor/processor.go]: moved validation to top

File: router/router.go [hunk 12]
```diff
@@ -1373,7 +1373,6 @@ func (rt *HandleT) commitStatusList(responseList *[]jobResponseT) {
 	}
 
 	// REPORTING - ROUTER - START
-	utilTypes.AssertSameKeys(connectionDetailsMap, statusDetailsMap)
```
[the remark this answers is not in the document]
Comment (Jayachand) [reply router/router.go:1376]: This validation doesnot seem to add much value. Addition of key to connectionDetailsMap is immediately followed by addition of same key to statusDetailsMap. Can be removed at all places.

File: processor/processor.go [hunk 13]
```diff
@@ -1914,6 +1994,11 @@ func (proc *Handle) transformSrcDest(
 		inCountMetadataMap = make(map[string]MetricMetadata)
 		for i := range eventList {
 			event := &eventList[i]
+			// enrich event metadata with transformation details if transformation is enabled
+			if transformationEnabled {
+				event.Metadata.TransformationId = destination.Transformations[0].ID
+				event.Metadata.TransformationVersionId = destination.Transformations[0].VersionID
+			}
```
[the remark this answers is not in the document]
Comment (Jayachand) [reply processor/processor.go]: done

File: processor/processor.go [hunk 14]
```diff
@@ -910,17 +914,43 @@ func (proc *Handle) recordEventDeliveryStatus(jobsByDestID map[string][]*jobsdb.
...
 	for i := range response.Events {
 		// Update metrics maps
 		userTransformedEvent := &response.Events[i]
-		proc.updateMetricMaps(successCountMetadataMap, successCountMap, connectionDetailsMap, statusDetailsMap, userTransformedEvent, jobsdb.Succeeded.State, func() json.RawMessage { return []byte(`{}`) })
+		var messages []types.SingularEventT
+		if len(userTransformedEvent.Metadata.MessageIDs) > 0 {
+			messageIds := userTransformedEvent.Metadata.MessageIDs
+			for _, msgID := range messageIds {
+				messages = append(messages, eventsByMessageID[msgID].SingularEvent)
+			}
```
Comment (atzoum) [inline processor/processor.go:932]: This could become a one-liner (I also don't see any need to do `append`)
```suggestion
			messages = lo.Map(userTransformedEvent.Metadata.MessageIDs, func(msgID string, _ int) types.SingularEventT { return eventsByMessageID[msgID].SingularEvent })
```

File: processor/processor.go [hunk 15]
```diff
@@ -910,17 +914,43 @@ func (proc *Handle) recordEventDeliveryStatus(jobsByDestID map[string][]*jobsdb.
...
+		} else {
+			messages = append(messages, eventsByMessageID[userTransformedEvent.Metadata.MessageID].SingularEvent)
```
Comment (atzoum) [inline processor/processor.go:934]: I don't see the need to append here either
```suggestion
			messages = []types.SingularEventT{eventsByMessageID[userTransformedEvent.Metadata.MessageID].SingularEvent}
```

File: processor/processor.go [hunk 16]
```diff
@@ -910,17 +914,43 @@ func (proc *Handle) recordEventDeliveryStatus(jobsByDestID map[string][]*jobsdb.
...
+		}
+
+		for _, message := range messages {
```
Comment (atzoum) [inline processor/processor.go:937]: **Question:** Why are we repeating this for every message? The only thing the message affects is the sample payload. Would it be more efficient to increment counters in bulk and use a random sample?

File: processor/processor.go [hunk 17]
```diff
@@ -1289,10 +1370,10 @@ func (proc *Handle) processJobsForDest(partition string, subJobs subJob, parsedE
 	inCountMap := make(map[string]int64)
 	inCountMetadataMap := make(map[string]MetricMetadata)
 	connectionDetailsMap := make(map[string]*types.ConnectionDetails)
-	statusDetailsMap := make(map[string]*types.StatusDetail)
+	statusDetailsMap := make(map[string]map[string]*types.StatusDetail)
```
Comment (atzoum) [inline processor/processor.go:1373]: What kind of key does this second map level holds? I find it hard to understand

File: utils/types/reporting_types.go [hunk 18]
```diff
@@ -35,6 +35,14 @@ var (
 	WAREHOUSE              = "warehouse"
 )
 
+var (
+	// Tracking plan validation states
+	SUCCEEDED                    = "succeeded"
+	SUCCEEDED_WITHOUT_VIOLATIONS = "succeeded_without_violations"
```
[the remark this answers is not in the document]
Comment (Jayachand) [reply utils/types/reporting_types.go:41]: In trackingplan stage, there will be no status SUCCEEDED unlike other stages.
`SUCCEEDED` is split into `SUCCEEDED_WITHOUT_VIOLATIONS` + `SUCCEEDED_WITH_VIOLATIONS`

[the remark this answers is not in the document]
Comment (atzoum) [reply utils/types/reporting_types.go:41]: Can't we skip `SUCCEEDED_WITHOUT_VIOLATIONS`? If it succeeded, it succeeded. If there are violations, I would also skip the `SUCCEEDED` part and use something more straightforward, e.g. `VIOLATIONS_FOUND`, since what you are interested in is mostly that violations were found, not that we are ignoring these violations by considering the events as succeeded and letting them flow further in the pipeline.

File: enterprise/reporting/reporting.go [hunk 19]
```diff
@@ -279,22 +321,25 @@ func (*HandleT) getAggregatedReports(reports []*types.ReportByStatus) []*types.M
...
+				ViolationCount: report.StatusDetail.ViolationCount,
 				SampleResponse: report.StatusDetail.SampleResponse,
 				SampleEvent:    report.StatusDetail.SampleEvent,
 				EventName:      report.StatusDetail.EventName,
 				EventType:      report.StatusDetail.EventType,
+				ErrorType:      report.StatusDetail.ErrorType,
 			})
 			continue
 		}
 		statusDetail := statusDetailInterface.(*types.StatusDetail)
```
[the remark this answers is not in the document]
Comment (Jayachand) [reply enterprise/reporting/reporting.go]: done

File: utils/types/reporting_types.go [hunk 20]
```diff
@@ -133,7 +153,7 @@ func CreatePUDetails(inPU, pu string, terminalPU, initialPU bool) *PUDetails {
 	}
 }
 
-func AssertSameKeys(m1 map[string]*ConnectionDetails, m2 map[string]*StatusDetail) {
+func AssertSameKeys(m1 map[string]*ConnectionDetails, m2 map[string]map[string]*StatusDetail) {
```
[the remark this answers is not in the document]
Comment (Jayachand) [reply utils/types/reporting_types.go]: done

File: processor/processor.go (hunk 14 above)
Comment (Jayachand) [reply processor/processor.go]: done

File: processor/processor.go (hunk 15 above)
Comment (Jayachand) [reply processor/processor.go]: done. added in getfailedEvents too

File: utils/types/reporting_types.go [hunk 21]
```diff
@@ -35,6 +35,14 @@ var (
 	WAREHOUSE              = "warehouse"
 )
 
+var (
+	// Tracking plan validation states
+	SUCCEEDED                    = "succeeded"
+	SUCCEEDED_WITHOUT_VIOLATIONS = "succeeded_without_violations"
+	SUCCEEDED_WITH_VIOLATIONS    = "succeeded_with_violations"
+	ABORTED                      = "aborted"
```
[the remark this answers is not in the document]
Comment (Jayachand) [reply utils/types/reporting_types.go]: deleted

File: utils/types/reporting_types.go [hunk 22]
```diff
@@ -35,6 +35,14 @@ var (
 	WAREHOUSE              = "warehouse"
 )
 
+var (
+	// Tracking plan validation states
+	SUCCEEDED                    = "succeeded"
```
[the remark this answers is not in the document]
Comment (Jayachand) [reply utils/types/reporting_types.go]: removed

File: utils/types/reporting_types.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (Jayachand) [reply utils/types/reporting_types.go]: All events sent to tracking plan stage can be categorized into 3 disjoint sets `SUCCEEDED_WITHOUT_VIOLATIONS` + `SUCCEEDED_WITH_VIOLATIONS` + `ABORTED (due to violations)`. User is more interested in `VIOLATIONS_FOUND (SUCCEEDED_WITH_VIOLATIONS+ ABORTED)` and `ABORTED`.Also User might ask for only succeeded without violations. In view of this, 3 disjoint status data are sent to reporting, graphs shown in UI can be constructed based on user requirements

[the remark this answers is not in the document]
Comment (Jayachand) [reply utils/types/reporting_types.go]: Letting events succeed with violations is a configuration setting given to user. User might this setting always without dropping events.

File: router/batchrouter/batchrouter.go (hunk 7 above)
Comment (Jayachand) [reply router/batchrouter/batchrouter.go]: done

**Review by atzoum — APPROVED** [review]
Collection of reporting metrics through metric maps was already a complicated affair. Unfortunately, now it appears to be even more fragile, complex and hard to follow, comprehend or reason about.
With the current design, I fear that every time there is a need to capture even more granular statistics, the complexity will grow exponentially.

File: processor/processor.go [hunk 23]
```diff
@@ -1572,6 +1650,10 @@ func (proc *Handle) processJobsForDest(partition string, subJobs subJob, parsedE
 					// At the TP flow we are not having destination information, so adding it here.
 					shallowEventCopy.Metadata.DestinationID = destination.ID
 					shallowEventCopy.Metadata.DestinationType = destination.DestinationDefinition.Name
+					if len(destination.Transformations) > 0 {
```
[the remark this answers is not in the document]
Comment (Jayachand) [reply processor/processor.go:1653]: A destination can have at max one transformation

File: processor/processor.go (hunk 17 above)
Comment (Jayachand) [reply processor/processor.go:1385]: Yeah, complexity is ever growing. 
I think we should plan for complete refactor of reporting metrics
